src/main/java/org/interledger/ilp/exceptions/InterledgerException.java

```diff
@@ -1,36 +1,207 @@
+     * Check the RFC https://interledger.org/rfcs/0003-interledger-protocol/#errors
+     * for the newest updated doc.
+     * 
+     * @param errCode,    one of the defined InterledgerException.ErrorCode defined.
+     * @param triggeredBy ILP address (prefix if it's a ledger) of the entity that originally emitted the error
+     * @param forwardedBy ILP address list of connectors that relayed the error message
+     * @param triggeredAt Time when the error was initially emitted
+     * @param data        OCTET STRINGSIZE(0..8192) Unless otherwise specified, data SHOULD be encoded as UTF-8. 
+     *                    Protocols built on top of ILP SHOULD specify the encoding format of error data 
+     */
+    public InterledgerException(
+        ErrorCode errCode,
+        InterledgerAddress triggeredBy,
+        ArrayList<InterledgerAddress> forwardedBy,
+        ZonedDateTime triggeredAt, 
+        String data) {
+        super();
+        if (errCode == null || triggeredBy == null || forwardedBy == null || data == null) {
+            throw new IllegalArgumentException("parameters can NOT be null");
+        }
+        this.errCode     = errCode;
+        this.triggeredBy = triggeredBy;
+        this.forwardedBy = forwardedBy;
+        this.triggeredAt = triggeredAt;
+        this.data        = data;
+    }
+    
+    /**
+     * Constructs an instance of <code>InterledgerException</code>
+     * with default parameters for forwardedBy (Empty list) and 
+     * triggeredAt (ZonedDateTime.now()). 
+     *  In most situations such values match the default ones 
+     * when triggering a new exception (vs an exception received
+     * from another ILP node that is being forwarded back to 
+     * originating request clients)
+     * 
+     * Check the RFC https://interledger.org/rfcs/0003-interledger-protocol/#errors
+     * for the newest updated doc.
+     */
+    public InterledgerException(
+        ErrorCode errCode,
+        InterledgerAddress triggeredBy,
+        String data) {
+        this(errCode, triggeredBy, EMPTY_ForwardedBy_List, ZonedDateTime.now(), data);
+    }
+
+    public ErrorCode getErrCode() {
+        return errCode;
+    }
+
+    /**
+     * @return ErrorType categorizes the error as FINAL, TEMPORARY or RELATIVE
+     *         (See RFC for more info)
+     */
+    public ErrorType getErrorType(){
+        return errCode.errorType;
+    }
+
+    public InterledgerAddress getTriggeredBy() {
+        return triggeredBy;
+    }
+
+    public ArrayList<InterledgerAddress> getForwardedBy() {
+        return forwardedBy;
+    }
+
+    public ZonedDateTime getTriggeredAt() {
+        return triggeredAt;
+    }
+
+    public String getData() {
+        return data;
+    }
+
+    /**
+     * Used by connectors to add themself to the forwaredBy list
+     * @param selfAdd connector ILP address.
+     */
+    public void addSelfToForwardedByList(InterledgerAddress selfAdd) {
```

Comment:
Perhaps add a note here to indicate that this is _not_ threadsafe?

From the Javadoc of ArrayList:

> <p><strong>Note that this implementation is not synchronized.</strong>
> If multiple threads access an <tt>ArrayList</tt> instance concurrently,
> and at least one of the threads modifies the list structurally, it
> <i>must</i> be synchronized externally.  (A structural modification is
> any operation that adds or deletes one or more elements, or explicitly
> resizes the backing array; merely setting the value of an element is not
> a structural modification.)  This is typically accomplished by
> synchronizing on some object that naturally encapsulates the list.
>
> If no such object exists, the list should be "wrapped" using the
> {@link Collections#synchronizedList Collections.synchronizedList}
> method.  This is best done at creation time, to prevent accidental
> unsynchronized access to the list:
> <pre>List list = Collections.synchronizedList(new ArrayList(...));</pre>

...Or we could wrap the EMPTY_ForwardedBy_List at constructions time?